File: nodejs/packages/node-red-contrib-periph-io-expander/nodes/tpic6b595/tpic6b595.js
'use strict';

module.exports = function(RED) {
    const opengpio = require('opengpio');
    const { SiPoConnection } = require('periph/src/connection/sipo');
    const { Tpic6b595Full }   = require('periph/src/chips/io_expander/tpic6b595');

    function Tpic6b595Node(config) {
        RED.nodes.createNode(this, config);
        const node = this;
        try {
            const rck   = new opengpio.Output(parseInt(config.rckPin,  10));
            const srclr = (config.srclrPin !== undefined && config.srclrPin !== '' && parseInt(config.srclrPin, 10) >= 0)
                            ? new opengpio.Output(parseInt(config.srclrPin, 10)) : null;
            const g     = (config.gPin !== undefined && config.gPin !== '' && parseInt(config.gPin, 10) >= 0)
                            ? new opengpio.Output(parseInt(config.gPin, 10))     : null;
            const serIn = config.serInPin ? new opengpio.Output(parseInt(config.serInPin, 10)) : null;
            const srck  = config.srckPin  ? new opengpio.Output(parseInt(config.srckPin,  10)) : null;

            const opts = { srclr, g };
            if (serIn && srck) { opts.serIn = serIn; opts.srck = srck; }
            const connection = new SiPoConnection(rck, opts);
            node.connection  = connection;
            node.chip        = new Tpic6b595Full(connection, parseInt(config.numDevices, 10) || 1);
        } catch (e) {
            node.error('TPIC6B595 init failed: ' + e.message);
        }

        node.on('input', async function(msg, send, done) {
            if (!node.chip) { done(new Error('TPIC6B595 not initialised')); return; }
            try {
                const p = msg.payload;
                if (p && typeof p.pin === 'number' && typeof p.value === 'number') {
                    await node.chip.pin(p.pin).write(p.value ? 1 : 0);
                    done();
                } else if (p && typeof p.port === 'number' && typeof p.value === 'number') {
                    await node.chip.writePort(p.port, p.value & 0xFF);
                    done();
                } else if (p && p.command === 'off') {
                    await node.chip.off();
                    done();
                } else if (p && typeof p.outputEnable === 'boolean') {
                    node.chip.setOutputEnable(p.outputEnable);
                    done();
                } else if (p && Array.isArray(p.values)) {
                    node.chip.writeAll(p.values);
                    done();
                } else {
                    done(new Error('msg.payload must be {pin,value}, {port,value}, {command:"off"}, {outputEnable:true|false}, or {values:[...]}.'));
                }
            } catch (e) { done(e); }
        });

        node.on('close', async function() {
            if (node.connection) await node.connection.close();
        });
    }

    RED.nodes.registerType('periph-tpic6b595', Tpic6b595Node);
};

File: python/uiflow2/io_expander/tpic6b595/Tpic6b595.py
"""
file     Tpic6b595
time     2026-09-21
author
email
license  Apache License 2.0
"""

from machine import SPI, Pin
from periph.connection.sipo_micropython import SiPoConnection
from periph.chips.io_expander.tpic6b595 import Tpic6b595Full


class Tpic6b595:
    """
    note:
        en: ''
    details:
        color: '#C084FC'
        link: https://github.com/tuhde/Periph
        image: ''
        category: Custom
    example: ''
    """

    def __init__(self, bus: int = 0, device: int = 0, rck: int = 17,
                 srclr: int = -1, g: int = -1, num_devices: int = 1):
        """
        label:
            en: '%1 init bus %2 device %3 rck %4 srclr %5 g %6 num_devices %7'
        params:
            bus:
                name: bus
                type: int
                default: '0'
                field: number
            device:
                name: device
                type: int
                default: '0'
                field: number
            rck:
                name: rck
                type: int
                default: '17'
                field: number
            srclr:
                name: srclr
                type: int
                default: '-1'
                field: number
            g:
                name: g
                type: int
                default: '-1'
                field: number
            num_devices:
                name: num_devices
                type: int
                default: '1'
                field: number
        """
        _spi = SPI(bus, baudrate=1000000, polarity=0, phase=0,
                   sck=Pin(rck + 1), mosi=Pin(rck + 2))
        _rck_pin = Pin(rck, Pin.OUT)
        _srclr_pin = Pin(srclr, Pin.OUT) if srclr >= 0 else None
        _g_pin = Pin(g, Pin.OUT) if g >= 0 else None
        _connection = SiPoConnection(_spi, _rck_pin, srclr=_srclr_pin, g=_g_pin)
        self._driver = Tpic6b595Full(_connection, num_devices=num_devices)

    def write_pin(self, pin: int, value: int):
        """
        label:
            en: '%1 write pin value (0/1) pin %2 value %3'
        params:
            pin:
                name: pin
                type: int
                field: number
            value:
                name: value
                type: int
                field: number
        """
        self._driver.pin(pin).value(bool(value))

    def write_port(self, port: int, mask: int):
        """
        label:
            en: '%1 write device byte (port) %2 mask (0x00-0xFF) %3'
        params:
            port:
                name: port
                type: int
                field: number
            mask:
                name: mask
                type: int
                field: number
        """
        self._driver.write_port(port, mask)

    def fill(self, value: int):
        """
        label:
            en: '%1 fill all outputs (0/1) value %2'
        params:
            value:
                name: value
                type: int
                field: number
        """
        self._driver.fill(bool(value))

    def off(self):
        """
        label:
            en: '%1 turn every output off'
        """
        self._driver.off()

    def clear(self):
        """
        label:
            en: '%1 pulse SRCLR (shift-register clear)'
        """
        self._driver.clear()

    def set_output_enable(self, enabled: int):
        """
        label:
            en: '%1 set output enable (0/1) via G enabled %2'
        params:
            enabled:
                name: enabled
                type: int
                field: number
        """
        self._driver.set_output_enable(bool(enabled))

    def write_all(self, d0: int, d1: int):
        """
        label:
            en: '%1 write all device bytes [d0 %2 d1 %3]'
        params:
            d0:
                name: d0
                type: int
                field: number
            d1:
                name: d1
                type: int
                field: number
        """
        self._driver.write_all([d0, d1])
